### PR TITLE
Fixes #1472: Removed the adaptor_finalizing flag. Moved connection fr…

### LINKED ISSUE #1472: Remove the adaptor_finalizing flag from the tcp_adaptor
Modify the delete connector and listener functions to not loop thru the connections and free them. Instead these delete functions must free the listener/connector only if there are no associated connections.

Move the connection/free loop to ADAPTOR_final() final

### PR DESCRIPTION
**Pull request by ganeshmurthy**
…ee loop to ADAPTOR_final().

Also introduced connector and listener incref and decref functions to be able to safely free connectors and listeners

### CODE REVIEW

File: src/adaptors/tcp/tcp_adaptor.c [hunk 1]
```diff
@@ -2532,16 +2534,26 @@ static void ADAPTOR_final(void *adaptor_context)
...
-        qd_dispatch_delete_tcp_connector(tcp_context->qd, cr);
+    while (DEQ_HEAD(tcp_context->listeners)) {
+        qd_tcp_listener_t *listener   = DEQ_HEAD(tcp_context->listeners);
+        qd_tcp_connection_t *conn = DEQ_HEAD(listener->connections);
+        while (conn) {
+            qd_tcp_connection_t *next_conn = DEQ_NEXT(conn);
+            close_connection_XSIDE_IO(conn);
+            conn = next_conn;
+        }
+        qd_tcp_listener_decref(listener);
```
Comment (kgiusti) [inline src/adaptors/tcp/tcp_adaptor.c:2545]: This call to decref may crash if the last closed tcp conn is the last entity holding the listener decref - in that case close_connection_XSIDE_IO will drop the refcount to zero and free the listener.

Same issue applies to the decref of the connector as well.

Fixing that may be tricky...

Without too much "brain smart-making" you could try the following hack:

1. modify the _decref() functions to return "true" if the listener/connector was freed else false
2. at the top of the while loop calll _incref before closing any connections. That ensures that the listener/connector won't be freed by any of the connections closing
3. after closing connections, loop calling _decref() until true is returned.

That's pretty hacky.  You could also fetch the refcount before freeing conns and only call _decref if the total number of conns (DEQ_SIZE) is less than the original refcount.

Hackity hack hack!

**Review by kgiusti — CHANGES_REQUESTED** [review]
Some comment's don't apply to this patch and can be handled in a follow up patch.

Of all the comments I think the ADAPTOR_final cleanup refcounting needs to be addressed (assuming I'm correct regarding the possible double free).

File: src/adaptors/tcp/tcp_adaptor.c [hunk 2]
```diff
@@ -2237,93 +2257,109 @@ static const char *tcp_alpn_protocols[TCP_NUM_ALPN_PROTOCOLS] = {"http/1.1", "h2
...
+        // Explicitly drop the out-link so that we notify any link event monitors and stop new deliveries from being
+        // forwarded to this connector
+        //
+        if (!!connector->out_link) {
+            qdr_link_set_context(connector->out_link, 0);
+            qdr_link_detach(connector->out_link, QD_LOST, 0);
+            connector->out_link = 0;
         }
+
+        qdr_connection_closed(connector->core_conn);
```
[the remark this answers is not in the document]
Comment (ganeshmurthy) [reply src/adaptors/tcp/tcp_adaptor.c:2356]: Done ! Thanks.

File: src/adaptors/tcp/tcp_adaptor.c [hunk 3]
```diff
@@ -459,27 +472,23 @@ static void free_connection_IO(void *context)
...
-            free_listener(li);
-        }
+        qd_tcp_listener_t *listener = (qd_tcp_listener_t*) conn->common.parent;
+        sys_mutex_lock(&listener->lock);
+        DEQ_REMOVE(listener->connections, conn);
+        sys_mutex_unlock(&listener->lock);
+        //
+        // Call listener decref when a connection associated with the listener is removed (DEQ_REMOVE(listener->connections, conn))
+        //
+        qd_tcp_listener_decref(listener);
```
[the remark this answers is not in the document]
Comment (ganeshmurthy) [reply src/adaptors/tcp/tcp_adaptor.c:482]: Done !

File: src/adaptors/tcp/tcp_adaptor.c [hunk 4]
```diff
@@ -459,27 +472,23 @@ static void free_connection_IO(void *context)
     // Do NOT free the core_activation lock since the core may be holding it
 
     if (!!conn->common.parent && conn->common.parent->context_type == TL_LISTENER) {
```
[the remark this answers is not in the document]
Comment (ganeshmurthy) [reply src/adaptors/tcp/tcp_adaptor.c:474]: Done !

File: src/adaptors/tcp/tcp_adaptor.c [hunk 5]
```diff
@@ -1182,6 +1191,11 @@ static uint64_t handle_first_outbound_delivery_CSIDE(qd_tcp_connector_t *connect
 
     sys_mutex_lock(&connector->lock);
     DEQ_INSERT_TAIL(connector->connections, conn);
+    //
+    // Call connector incref when a connection associated with the connector is added (DEQ_INSERT_TAIL(connector->connections, conn))
```
[the remark this answers is not in the document]
Comment (ganeshmurthy) [reply src/adaptors/tcp/tcp_adaptor.c:1195]: Done !

File: src/adaptors/tcp/tcp_adaptor.c [hunk 6]
```diff
@@ -1947,34 +1961,40 @@ static void on_connection_event_CSIDE_IO(pn_event_t *e, qd_server_t *qd_server,
...
 
-    sys_mutex_lock(&li->lock);
-    DEQ_INSERT_TAIL(li->connections, conn);
-    li->connections_opened++;
-    vflow_set_uint64(li->common.vflow, VFLOW_ATTRIBUTE_FLOW_COUNT_L4, li->connections_opened);
-    sys_mutex_unlock(&li->lock);
+    sys_mutex_lock(&listener->lock);
+    DEQ_INSERT_TAIL(listener->connections, conn);
+    //
+    // Call listener incref when a connection associated with the listener is added (DEQ_INSERT_TAIL(listener->connections, conn))
```
[the remark this answers is not in the document]
Comment (ganeshmurthy) [reply src/adaptors/tcp/tcp_adaptor.c:1991]: Done !

**Review by kgiusti — APPROVED** [review]
Excellent! Ship it.

**Pull request merged**